heron/ckptmgr/src/java/com/twitter/heron/ckptmgr/CheckpointManager.java

```diff
@@ -0,0 +1,247 @@
+  private static Options constructOptions() {
+    Options options = new Options();
+
+    Option topName = Option.builder("t")
+        .desc("Name of the topology")
+        .longOpt("topologyname")
+        .hasArgs()
+        .argName("topologyname")
+        .required()
+        .build();
+
+    Option topId = Option.builder("i")
+        .desc("Id of the topology")
+        .longOpt("topologyid")
+        .hasArgs()
+        .argName("topologyid")
+        .required()
+        .build();
+
+    Option ckptMgrId = Option.builder("c")
+        .desc("Id of the checkpoint manager")
+        .longOpt("ckptmgrid")
+        .hasArgs()
+        .argName("ckptmgrid")
+        .required()
+        .build();
+
+    Option ckptMgrPort = Option.builder("p")
+        .desc("Port of the checkpoint manager")
+        .longOpt("ckptmgrport")
+        .hasArgs()
+        .argName("ckptmgrport")
+        .required()
+        .build();
+
+    Option ckptMgrConfig = Option.builder("f")
+        .desc("Config name of the checkpoint manager")
+        .longOpt("ckptmgrconfig")
+        .hasArgs()
+        .argName("ckptmgrconfig")
+        .required()
+        .build();
+
+    Option heronInternalConfig = Option.builder("g")
+        .desc("Heron internal config filename")
+        .longOpt("heroninternalconfig")
+        .hasArgs()
+        .argName("heroninternalconfig")
+        .required()
+        .build();
+
+    options.addOption(topName);
+    options.addOption(topId);
+    options.addOption(ckptMgrId);
+    options.addOption(ckptMgrPort);
+    options.addOption(ckptMgrConfig);
+    options.addOption(heronInternalConfig);
+
+    return options;
+  }
+
+  // construct command line help options
+  private static Options constructHelpOptions() {
+    Options options = new Options();
+    Option help = Option.builder("h")
+        .desc("List all options and their description")
+        .longOpt("help")
+        .build();
+
+    options.addOption(help);
+    return options;
+  }
+
+  public CheckpointManager(
+      String topologyName, String topologyId, String checkpointMgrId,
+      String serverHost, int serverPort,
+      SystemConfig systemConfig, CheckpointManagerConfig checkpointManagerConfig)
+      throws IOException, CheckpointManagerException {
+
+    this.checkpointManagerServerLoop = new NIOLooper();
+
+    HeronSocketOptions serverSocketOptions =
+        new HeronSocketOptions(
+            checkpointManagerConfig.getWriteBatchSizeBytes(),
+            checkpointManagerConfig.getWriteBatchTimeMs(),
+            checkpointManagerConfig.getReadBatchSizeBytes(),
+            checkpointManagerConfig.getReadBatchTimeMs(),
+            checkpointManagerConfig.getSocketSendSize(),
+            checkpointManagerConfig.getSocketReceiveSize());
+
+    // Setup the IStatefulStorage
+    // TODO(mfu): This should be done in an executor driven by another thread, kind of async
+    IStatefulStorage checkpointsBackend;
```

Comment:
WRT refactor of IStatefulStorage, we really cannot predict what kind of storage mechanism will be used. The first two might look similar but lets say if the next one is cassandra, then no longer we deal with files/dirs but with keyspaces.